### ISSUE TITLE
[typesystem] Strange validation error

### ISSUE THREAD

**Issue by JanKoehnlein** [eclipse/xtext-xtend#7]
The following Xtend method yields a validation error

```
def <T> int foo(T x) {
  if(x instanceof String)  // Incompatible conditional operand types T and String
     return 1;
  else
    return 2;
}
```

The Java equivalent compiles fine.

**Comment by spoenemann** [comment]
This is probably on Xbase level?

**Comment by jfgossage** [comment]
I just checked this out and on Windows 10 using both Oracle Java 8 and Java 10, Java gives the same error as Xtend. Can anyone reproduce this?

**Comment by cdietrich** [comment]
@jfgossage i am not getting any error on a pure java file. can you give some hints how you have tried this?

**Comment by szarnekow** [comment]
You'd get an error, if T is proven to be no string, e.g. `T extends Number`. In any case: the validation is broken.

**Comment by jfgossage** [comment]
@cdietrich Unfortunately I did not keep my test case but when I tried to reproduce it again I get the following results for this Java class:

```
public class Bug7 {
	public <Number> int foo(Number x) {
		if (x instanceof String) {
			return 1;
		} else {
			return 2;
		}
	}
}
```

This gives a warning `The type parameter Number is hiding the type Number`.

If I change the Java class to:

```
public class Bug7 {
	public <Number> int foo(Integer x) {
		if (x instanceof String) {
			return 1;
		} else {
			return 2;
		}
	}
}
```

I do get the error `Incompatible conditional operand types Integer and String`.

**Comment by cdietrich** [comment]
Fixed in 2.17
